criu/util.c

```diff
@@ -1640,3 +1641,207 @@ ssize_t write_all(int fd, const void *buf, size_t size)
 	}
 	return n;
 }
+
+int rm_rf(char *target)
+{
+	int offset = strlen(target);
+	DIR *dir = NULL;
+	struct dirent *de;
+	int ret = -1;
+
+	dir = opendir(target);
+	if (!dir) {
+		pr_perror("unable to open %s", target);
+		return -1;
+	}
+
+	while ((de = readdir(dir))) {
+		int n;
+
+		if (dir_dots(de))
+			continue;
+
+		n = snprintf(target+offset, PATH_MAX-offset, "/%s", de->d_name);
+		if (n < 0 || n >= PATH_MAX) {
+			pr_err("snprintf failed\n");
+			goto out;
+		}
+
+		if (de->d_type == DT_DIR && rm_rf(target))
+			goto out;
+
+		if (remove(target) < 0) {
+			pr_perror("unable to remove %s", target);
+			goto out;
+		}
+	}
+
+	ret = 0;
+out:
+	target[offset] = 0;
+	return ret;
+}
+
+__attribute__((returns_twice)) static pid_t raw_legacy_clone(unsigned long flags, int *pidfd)
+{
+#if defined(__s390x__) || defined(__s390__) || defined(__CRIS__)
+	/* On s390/s390x and cris the order of the first and second arguments
+	 * of the system call is reversed.
+	 */
+	return syscall(__NR_clone, NULL, flags | SIGCHLD, pidfd);
+#elif defined(__sparc__) && defined(__arch64__)
+	{
+		/*
+		 * sparc64 always returns the other process id in %o0, and a
+		 * boolean flag whether this is the child or the parent in %o1.
+		 * Inline assembly is needed to get the flag returned in %o1.
+		 */
+		register long g1 asm("g1") = __NR_clone;
+		register long o0 asm("o0") = flags | SIGCHLD;
+		register long o1 asm("o1") = 0; /* is parent/child indicator */
+		register long o2 asm("o2") = (unsigned long)pidfd;
+		long is_error, retval, in_child;
+		pid_t child_pid;
+
+		asm volatile(
+#if defined(__arch64__)
+		    "t 0x6d\n\t" /* 64-bit trap */
+#else
+		    "t 0x10\n\t" /* 32-bit trap */
+#endif
+		    /*
+		     * catch errors: On sparc, the carry bit (csr) in the
+		     * processor status register (psr) is used instead of a
+		     * full register.
+		     */
+		    "addx %%g0, 0, %%g1"
+		    : "=r"(g1), "=r"(o0), "=r"(o1), "=r"(o2) /* outputs */
+		    : "r"(g1), "r"(o0), "r"(o1), "r"(o2)     /* inputs */
+		    : "%cc");				     /* clobbers */
+
+		is_error = g1;
+		retval = o0;
+		in_child = o1;
+
+		if (is_error) {
+			errno = retval;
+			return -1;
+		}
+
+		if (in_child)
+			return 0;
+
+		child_pid = retval;
+		return child_pid;
+	}
+#elif defined(__ia64__)
+	/* On ia64 the stack and stack size are passed as separate arguments. */
+	return syscall(__NR_clone, flags | SIGCHLD, NULL, 0, pidfd);
+#else
+	return syscall(__NR_clone, flags | SIGCHLD, NULL, pidfd);
+#endif
+}
+
+__attribute__((returns_twice)) static pid_t raw_clone(unsigned long flags, int *pidfd)
+{
+	pid_t pid;
+	struct _clone_args args = {
+		.flags		= flags,
```

Comment:
interesting, in the other file you didn't align the struct initialization. but, that is mixed anyway all over the code base. so no need to change.